pkg/solana/txm/txm.go

```diff
@@ -377,135 +382,220 @@ func (txm *Txm) confirm() {
 		}
-		tick = time.After(utils.WithJitter(txm.cfg.ConfirmPollPeriod()))
+	}
+}
+
+// handleConfirmedSignatureStatus handles the case where a transaction signature is in the "confirmed" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleConfirmedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnConfirmed(sig)
+	if err != nil && !errors.Is(err, ErrAlreadyInExpectedState) {
+		txm.lggr.Errorw("failed to mark transaction as confirmed", "id", id, "signature", sig, "error", err)
+	} else if err == nil {
+		txm.lggr.Debugw("marking transaction as confirmed", "id", id, "signature", sig)
+	}
+}
+
+// handleFinalizedSignatureStatus handles the case where a transaction signature is in the "finalized" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleFinalizedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnFinalized(sig, txm.cfg.TxRetentionTimeout())
+	if err != nil {
+		txm.lggr.Errorw("failed to mark transaction as finalized", "id", id, "signature", sig, "error", err)
+	} else {
+		txm.lggr.Debugw("marking transaction as finalized", "id", id, "signature", sig)
+	}
+}
+
+// rebroadcastExpiredTxs attempts to rebroadcast all transactions that are in broadcasted state and have expired.
+// An expired tx is one where it's blockhash lastValidBlockHeight is smaller than the current slot height.
+// If any error occurs during rebroadcast attempt, they are discarded, and the function continues with the next transaction.
+func (txm *Txm) rebroadcastExpiredTxs(ctx context.Context, client client.ReaderWriter) {
+	currHeight, err := client.SlotHeight(ctx)
+	if err != nil {
+		txm.lggr.Errorw("failed to get current slot height", "error", err)
+		return
+	}
+	blockhash, err := client.LatestBlockhash(ctx)
+	if err != nil {
+		txm.lggr.Errorw("failed to get blockhash", "error", err)
+		return
+	}
+	// Rebroadcast all expired txes
+	for _, tx := range txm.txs.ListAllExpiredBroadcastedTxs(currHeight) {
+		txm.lggr.Infow("transaction expired, rebroadcasting", "id", tx.id, "signature", tx.signatures)
+		if len(tx.signatures) == 0 { // prevent panic, shouldn't happen.
+			txm.lggr.Errorw("no signatures found for expired transaction", "id", tx.id)
+			continue
+		}
+		_, err := txm.txs.Remove(tx.signatures[0]) // only picking signature[0]. Remove func removes all related remaining signatures.
+		if err != nil {
+			txm.lggr.Errorw("failed to remove expired transaction", "id", tx.id, "error", err)
+			continue
+		}
+		// Overwrite blockhash and lastValidBlockHeight with latest values so that the transaction can be rebroadcasted an accepted.
+		tx.tx.Message.RecentBlockhash = blockhash.Value.Blockhash
+		rebroadcastTx := pendingTx{
+			tx:                   tx.tx,
+			id:                   tx.id,
+			rebroadcastCount:     tx.rebroadcastCount + 1,
+			lastValidBlockHeight: blockhash.Value.LastValidBlockHeight,
+		}
```

Comment:
I think `sendWithRetry` also expects `cfg` to be set in the message
```suggestion
		rebroadcastTx := pendingTx{
			tx:                   tx.tx,
			id:                   tx.id,
			cfg:                  tx.cfg,
			rebroadcastCount:     tx.rebroadcastCount + 1,
			lastValidBlockHeight: blockhash.Value.LastValidBlockHeight,
		}
```